### PR TITLE
Adding tests to cover the array checks in the core Utils

### LINKED ISSUE #2213: Array Checks in Microsoft.ML.Internal.Utilities.Utils are not tested
The array tests used as sanity checks are not tested, even though they have very particular expectations on behavior (different from their names imply, sometimes). We should add tests to cover them.

### PR DESCRIPTION
**Pull request by rogancarr**
This PR adds tests to cover the array checks in `Microsoft.ML.Internal.Utilities`.

Fixes #2213

### CODE REVIEW

File: src/Microsoft.ML.Core/Utilities/Utils.cs [hunk 1]
```diff
@@ -534,7 +534,15 @@ public static int[] GetRandomPermutation(Random rand, int size)
             return res;
         }
 
-        public static bool AreEqual(Single[] arr1, Single[] arr2)
+        public static void Shuffle<T>(Random rand, Span<T> rgv)
```
Comment (rogancarr) [inline src/Microsoft.ML.Core/Utilities/Utils.cs]: Sorting functions -- this was in the middle of a batch of `AreEqual` implementation. I moved it up near the rest of the randomization utils.

File: src/Microsoft.ML.Core/Utilities/Utils.cs [hunk 2]
```diff
@@ -551,7 +559,7 @@ public static bool AreEqual(Single[] arr1, Single[] arr2)
             return true;
         }
 
-        public static bool AreEqual(Double[] arr1, Double[] arr2)
+        public static bool AreEqual(double[] arr1, double[] arr2)
```
Comment (rogancarr) [inline src/Microsoft.ML.Core/Utilities/Utils.cs]: Normalized naming while I was here.

File: test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        {
+            // An increasing array
+            int[] x = Enumerable.Range(0, 10).ToArray();
+            Assert.True(Utils.IsIncreasing(0, x, 10));
+            // Check the lower bound
+            Assert.False(Utils.IsIncreasing(1, x, 10));
+            // The upper bound should be exclusive
+            Assert.False(Utils.IsIncreasing(0, x, 9));
+            // Any len should work
+            Assert.True(Utils.IsIncreasing(0, x, 5, 10));
```
Comment (singlis) [inline test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs]: What about if x contains duplicate numbers in a sorted order (like 0,1,1,2,3,)?

File: test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+            // A sorted (increasing) array
+            int[] x = Enumerable.Range(0, 10).ToArray();
+            int[] y = Enumerable.Range(0, 10).ToArray();
+            Assert.True(Utils.AreEqual(x, y));
+
+            // Not Equal
+            var x1Temp = x[1];
+            x[1] = x[0];
+            Assert.False(Utils.AreEqual(x, y));
+            x[1] = x1Temp;
```
Comment (singlis) [inline test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs]: Minor nit, each test you are resetting back to the original value but its at the end of the test so the reset isnt necessary.

File: test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            x[7] = x[9];
+            Assert.False(Utils.IsIncreasing(0, x, 10));
+            // Before the mismatched entry, it should be fine
+            Assert.True(Utils.IsIncreasing(0, x, 7, 10));
+            x[1] = x7Temp;
+        }
+
+        [Fact]
+        [TestCategory("Utilities")]
+        public void CheckIsAreEqualInt()
```
Comment (singlis) [inline test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs]: minor, I would update CheckAreEqual* or add more tests to test different lengths, change the ends of the array ([0] and [length]), and null values being passed in.

File: test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs (hunk 3 above)
Comment (rogancarr) [reply test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs]: Covered by the lines just below this one! (I refer to this as monotonically increasing sequences.)

Comment (singlis) [reply test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs]: oh thats true, ok!

---
In reply to: [250402110](https://github.com/dotnet/machinelearning/pull/2214#discussion_r250402110) [](ancestors = 250402110)

File: test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs (hunk 4 above)
Comment (rogancarr) [reply test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs]: Yes. I debated this. I went for this to help out anybody who may want to test more conditions by adding onto this function.

File: test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs (hunk 5 above)
Comment (rogancarr) [reply test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs]: Good call! Done.

File: test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    public class TestUtilities : BaseTestBaseline
+    {
+        public TestUtilities(ITestOutputHelper helper)
+            : base(helper)
+        {
+        }
+
+        [Fact]
+        [TestCategory("Utilities")]
+        public void CheckIsSortedInt()
```
Comment (artidoro) [inline test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs]: >CheckIsSortedInt [](start = 20, length = 16)

Could you check the behavior with null input?

File: test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+
+            // Not sorted
+            x[1] = x[6];
+            Assert.False(Utils.IsSorted(x));
+            x[1] = x1Temp;
+        }
+
+        [Fact]
+        [TestCategory("Utilities")]
+        public void CheckIsSortedFloat()
```
Comment (artidoro) [inline test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs]: >CheckIsSortedFloat [](start = 20, length = 18)

Same here. Could you check the behavior with null input?

File: test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+            // but we want to guarantee that NaNs aren't allowed in these arrays.
+            var x0Temp = x[0];
+            x[0] = float.NaN;
+            Assert.False(Utils.IsSorted(x));
+            x[0] = x0Temp;
+        }
+
+        [Fact]
+        [TestCategory("Utilities")]
+        public void CheckIsIncreasing()
```
Comment (artidoro) [inline test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs]: >CheckIsIncreasing [](start = 20, length = 17)

Same here. Could you check the behavior with null input?

Comment (rogancarr) [reply test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs]: Done!

File: test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs (hunk 7 above)
Comment (rogancarr) [reply test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs]: Done!

File: test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs (hunk 6 above)
Comment (rogancarr) [reply test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs]: Done!
